sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/FileStreamSource.scala

```diff
@@ -67,12 +62,33 @@ class FileStreamSource(
   }
 
   /**
+   * Set the metadata path. This method should be called before using [[FileStreamSource]].
+   */
+  def setMetadataPath(metadataPath: String): Unit = {
```

Comment:
> I'd really prefer to avoid the pattern of having a initialization that is separate from the constructor.

Same as above. We don't know `metadataPath` when `DataSource.createSource` is called in `DataFrameReader`.